---
title: "Untitled"
author: "Beatriz Pereira"
date: "2023-03-08"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load necessary libraries
```{r}
library(dplyr)
library(tidyverse)
library(reshape2)
library(naniar)
library(ggplot2)
```

# Load your data: 
## gang_et_al_genotypes
```{r}
fang <- read_tsv("https://github.com/EEOB-BioData/BCB546_Spring2023/raw/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")
```

## snp_position
```{r}
snp <- read_tsv("https://github.com/EEOB-BioData/BCB546_Spring2023/raw/main/assignments/UNIX_Assignment/snp_position.txt")
```

# --- Part 1a, Data inspection ---
## gang
```{r}
type_fang <- typeof(fang)
### fang is a "list", which is multiple, heterogeneous, vectors with 1 element each. 
```
 
```{r}
length_fang <- length(fang)
### There are 986 elements
```

```{r}
dim_fang <- dim(fang)
### there are 2,782 rows, and 986 columns 
```

```{r}
structure_fang <- str(fang)
### this is a "data.frame", with 2,782 observations and 986 variables. Variable titles are as follows: Sample_ID, JG_OTU, Group, and the listed genotypes. 
```

```{r}
class_fang <- sapply(fang, class)
### every column has class: "character" 
```

```{r}
groups_fangs <- table(fang$Group)
### We will be filtering out desired Groups, so it's good to know how many of each there are, so we can look back and verify we haven't lost data. 
```

## snp
```{r}
type_snp <- typeof(snp)
### snp is a "list", which is multiple, heterogeneous, vectors with 1 element each.
```

```{r}
length_snp <- length(snp)
### there are 15 elements.
```

```{r}
dim_snp <- dim(snp)
### there are 983 rows, and 15 columns.
```

```{r}
structure_snp <- str(snp)
### The variables in this file include details about the SNP_ID itself - beggining with SNP_ID, and ending with count_gene.
```

```{r}
class_snp <- sapply(snp, class)
### unlike gang, snp has a mixture of characters and integers. 
```

# --- Part 1b,Data Processing ---
## pull out your desired groups for Maize and Teosinte.
```{r}
Subsetted_Maize <- filter(fang, Group %in% c("ZMMIL","ZMMLR","ZMMMR"))
```
```{r}
Subsetted_Teosinte <- filter(fang, Group %in% c("ZMPBA","ZMPIL","ZMPJA"))
```

## Checking that you have the right data: 
```{r}
table_subsettedMaize <- table(Subsetted_Maize$Group)
table_subsettedTeosinte <- table(Subsetted_Teosinte$Group)
### Results match what is in our original file (fang). 
```

## Remove columns that we do not need, including "JG_OTU" and "Group": 
```{r}
Subsetted_Maize <- Subsetted_Maize[,!names(Subsetted_Maize) %in% c("JG_OTU", "Group")]
```
```{r}
Subsetted_Teosinte <- Subsetted_Teosinte[,!names(Subsetted_Teosinte) %in% c("JG_OTU", "Group")]
```

## Change Sample_ID to SNP_ID (first column)
```{r}
colnames(Subsetted_Maize) [1] <- "SNP_ID"
```
```{r}
colnames(Subsetted_Teosinte) [1] <- "SNP_ID"
```

## Transpose the genotype data so that the columns become rows
```{r}
Subsetted_Maize_Transposed <- t(Subsetted_Maize)
```
```{r}
Subsetted_Teosinte_Transposed <- t(Subsetted_Teosinte)
### This should be clearly visible if you take a look at the file.
```
 

## Before joining, we only want the first 3 columns of the snp file. Additionally, the snp file should become a tibble for further processing:
```{r}
snp_ready <- snp[ , c("SNP_ID", "Chromosome","Position")]       
snp_ready <- tibble::as_tibble(snp_ready)
```

## Now, to join the snp_ready file with the modified + transposed genotype files:
```{r}
snp_maize <- merge(snp_ready, Subsetted_Maize_Transposed, by.x = "SNP_ID", by.y = "row.names")
```
```{r}
snp_teosinte <- merge(snp_ready, Subsetted_Teosinte_Transposed, by.x = "SNP_ID", by.y = "row.names")
```

## both should be tibbles for upcoming processing by dplyr:
```{r}
snp_maize <- tibble::as_tibble(snp_maize)
snp_teosinte <- tibble::as_tibble(snp_teosinte)
```

# Maize 20 files - Processing: 
## 10 files, ordered by increasing Position, with missing data encoded with "?":
```{r}
### Make a new directory for decreasing files: 
dir.create('./increasing_maize')

### Loop for increasing, with ? for ?/?:
for ( i in 1:10) {
  Chromosome_maize <- filter(snp_maize,Chromosome %in% c(i))
  Chromosome_maize_inc <- Chromosome_maize[order(as.numeric(as.character(Chromosome_maize$Position))),]
  Chromosome_maize_inc[Chromosome_maize_inc =="?/?"] <- "?"
  write.table(Chromosome_maize_inc, file =
paste("./increasing_maize/increasing_maize",i), quote = F, sep = "\t")
}
```

## 10 files, ordered by decreasing Position, with missing data encoded with "-":
```{r}
### Make a new directory for decreasing files: 
dir.create('./decreasing_maize')

### Loop for decreasing, with - for ?/?:
for ( i in 1:10) {
  Chromosome_maize_dec <- Chromosome_maize[order(-as.numeric(as.character(Chromosome_maize$Position))),]
  Chromosome_maize_dec[Chromosome_maize_dec =="?/?"] <- "-"
  write.table(Chromosome_maize_dec, file =
paste("./decreasing_maize/decreasing_maize",i), quote = F, sep = "\t")
}
```

# Teosinte 20 files: 
## 10 files, ordered by increasing Position, with missing data encoded with "?":
```{r}
### Make a new directory for decreasing files: 
dir.create('./increasing_teosinte')

### Loop for increasing, with ? for ?/?:
for ( i in 1:10) {
  Chromosome_teosinte <- filter(snp_teosinte,Chromosome %in% c(i))
  Chromosome_teosinte_inc <- Chromosome_teosinte[order(as.numeric(as.character(Chromosome_teosinte$Position))),]
  Chromosome_teosinte_inc[Chromosome_teosinte_inc =="?/?"] <- "?"
  write.table(Chromosome_teosinte_inc, file =
paste("./increasing_teosinte/increasing_teosinte",i), quote = F, sep = "\t")
}    
```

## 10 files, ordered by decreasing Position, with missing data encoded with "-":
```{r}
### Make a new directory for decreasing files: 
dir.create('./decreasing_teosinte')

### Loop for decreasing, with - for ?/?:
for ( i in 1:10) {
  Chromosome_teosinte_dec <- Chromosome_teosinte[order(-as.numeric(as.character(Chromosome_teosinte$Position))),]
  Chromosome_teosinte_dec[Chromosome_teosinte_dec =="?/?"] <- "-"
  write.table(Chromosome_teosinte_dec, file =
paste("./decreasing_teosinte/decreasing_teosinte",i), quote = F, sep = "\t")
}
```

# --- Part 2: Visualization ---
## Distribution of SNPs on each chromosome & across (Maize and Teosinte combined): 
```{r}
### prep
Subsetted_Teo_Maize <- filter(fang, Group %in% c("ZMMIL","ZMMLR","ZMMMR","ZMPBA","ZMPIL","ZMPJA"))
Subsetted_Teo_Maize <- Subsetted_Teo_Maize[,!names(Subsetted_Teo_Maize) %in% "JG_OTU"]
colnames(Subsetted_Teo_Maize) [1] <- "SNP_ID"
Subsetted_Teo_Maize_T <- t(Subsetted_Teo_Maize)
snp_teo_maize <- merge(snp_ready, Subsetted_Teo_Maize_T, by.x = "SNP_ID", by.y = "row.names", all = TRUE)
snp_teo_maize %>% naniar::replace_with_na_all(condition = ~.x == "?/?")
snp_teo_maize <- snp_teo_maize %>% pivot_longer(starts_with("V"),
                             names_to='geno',
                             values_to='SNP')
### plot
ggplot (data = snp_teo_maize) + geom_bar(mapping = aes(x = as.factor(as.double(Chromosome)), fill = Chromosome)) +xlab(label = "Chromosome") + ylab(label = "SNPs") + ggtitle("Single Nucleotide Polymorphism per Chromosome") + scale_x_discrete(limit=c(1:10))

### Chromosome 1 has the most SNPs, while chromosome 10 appears to have the least. 
```

## Distribution of SNPs on each chromosome & across (Maize vs. Teosinte)
```{r}
### prep
fang_new <- fang[,!names(fang) %in% "JG_OTU"]
fang_new <- fang_new %>% pivot_longer(!c(Sample_ID, Group), names_to="SNP_ID", values_to= "allele")
Subsetted_Teo_Maize_2 <- filter(fang_new, Group %in% c("ZMMIL","ZMMLR","ZMMMR","ZMPBA","ZMPIL","ZMPJA"))
snp_teo_maize_2 <- merge(Subsetted_Teo_Maize_2, snp_ready, by="SNP_ID")
snp_teo_maize_2 <- tibble::as_tibble(snp_teo_maize_2)
snp_teo_maize_groups <- snp_teo_maize_2 %>% 
  mutate(
    type = if_else(
      condition = Group %in% c("ZMMIL","ZMMLR","ZMMMR"), 
      true      = "Maize", 
      false     = "Teosinte"
    ), 
  )
### graph 1
by_group_plot <- ggplot (data = snp_teo_maize_groups) + geom_bar(mapping = aes(x = Chromosome, fill = type)) + 
xlab(label = "Chromosome") + ylab(label = "SNPs") +
  ggtitle("Single Nucleotide Polymorphism per Sample") 
by_group_plot

### graph 2
by_group_plot + facet_grid(Group ~ .)

### The group with the highest SNPs is the ZMMLR, belonging to Maize. Teosinte has an overall lower SNP count. 
```

## homozygous vs. heterozygous:
```{r}
### Labeling homozygous and heterozygous as TRUE (homozygous) or FALSE (heterozygous):
snp_homo_het <- merge(Subsetted_Teo_Maize_2, snp_ready, by="SNP_ID")
snp_homo_het <- tibble::as_tibble(snp_homo_het)
snp_homo_het <- snp_homo_het %>% replace_with_na(replace = list(allele = "?/?"))
snp_homo_het <- snp_homo_het %>% 
  mutate(
    homozygosity = if_else(
      condition = allele %in% c("A/A","T/T","G/G","C/C"), 
      true      = "homozygous", 
      false     = "heterozygous"
    ), 
  )
snp_homo_het <- snp_homo_het %>% 
  mutate(
    type = if_else(
      condition = Group %in% c("ZMMIL","ZMMLR","ZMMMR","C/C"), 
      true      = "Maize", 
      false     = "Teosinte"
    ), 
  )
snp_homo_het_na_omit <- na.omit(snp_homo_het) 
### plot het vs. hom in each group
ggplot(data = snp_homo_het_na_omit) + 
  geom_bar(mapping = aes(x = homozygosity, fill = Group)) + 
  xlab(label = "Homozygous vs. Heterozygous") + ylab(label = "Count") +
  ggtitle("Homozygous vs. Heterozygous by Group & Sample") + facet_grid(type ~ .) 

ggplot(data = snp_homo_het_na_omit) + 
  geom_bar(mapping = aes(x = Group, fill = homozygosity)) + 
  xlab(label = "Homozygous vs. Heterozygous") + ylab(label = "Count") +
  ggtitle("Homozygous vs. Heterozygous by Group") 

ggplot(data = snp_homo_het_na_omit) + 
  geom_bar(mapping = aes(x = type, fill = homozygosity)) + 
  xlab(label = "Homozygous vs. Heterozygous") + ylab(label = "Count") +
  ggtitle("Homozygous vs. Heterozygous in Maize + Teosinte")

### There is a close amount of heterozygosity in both samples, but we need to keep in mind that the Teosinte sample has a lower number of SNPs overall, and therefore, in my opinion, a conclusion cannot be drawn from this data. 
```

## NA data: 
```{r}
r_is_hard <- snp_homo_het %>% 
  mutate(
    missing_data = if_else(
      condition = allele %in% NA, 
      true      = "missing", 
      false     = "not missing"
    ), 
  )

ggplot(data = r_is_hard) + 
  geom_bar(mapping = aes(x = missing_data, fill = Group)) + 
  xlab(label = "Missing or Not") + ylab(label = "Count") +
  ggtitle("Missing Data in Maize + Teosinte") + facet_grid(type ~ .)

ggplot(data = r_is_hard) +
  geom_bar(mapping = aes(x = type, fill = missing_data)) + 
  xlab(label = "Sample") + ylab(label = "Count") +
  ggtitle("Missing Data in Maize + Teosinte")

### Proportionally, there is not a lot missing from teosinte, but SNP-wise, it's still concerning. 
```

## Creative portion of assignment (additional graph):
```{r}
### Number amplicons in samples
snp_ready_2 <- snp[ , c("SNP_ID", "Chromosome","Position","count_amplicons")]       
snp_ready_2 <- tibble::as_tibble(snp_ready_2)
snp_vis_2_new <- merge(Subsetted_Teo_Maize_2, snp_ready_2, by="SNP_ID")
snp_vis_2_new <- tibble::as_tibble(snp_vis_2_new)
Subsetted_SNPs_new <- snp_vis_2_new %>% 
  mutate(
    type = if_else(
      condition = Group %in% c("ZMMIL","ZMMLR","ZMMMR"), 
      true      = "Maize", 
      false     = "Teosinte"
    ), 
  )

ggplot(Subsetted_SNPs_new) + geom_col(aes(x = type, y = count_amplicons)) + xlab(label = "Sample") + ylab(label = "Number of Amplicons") + ggtitle("Amplicons per Sample")

### This difference in number of amplicons suggests that something went wrong during PCR, or a different step in the procedure with the Teosinte sample. 
```

